### ISSUE TITLE
EvlParallelOperationsTests crashes under Apple Silicon

### ISSUE THREAD

**Issue by agarciadom** [eclipse/epsilon#111]
In an Apple Silicon system (particularly, an M3 Pro-based laptop), `EvlParallelOperationsTests` (executed by our `build-and-run.sh` script) crashes with this trace:

```
java.lang.OutOfMemoryError: unable to create new native thread
	at java.lang.Thread.start0(Native Method)
	at java.lang.Thread.start(Thread.java:719)
	at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:957)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1367)
	at java.util.concurrent.AbstractExecutorService.invokeAll(AbstractExecutorService.java:238)
	at org.eclipse.epsilon.eol.execute.context.concurrent.IEolContextParallel.executeAll(IEolContextParallel.java:149)
	at org.eclipse.epsilon.eol.execute.operations.declarative.concurrent.ParallelSelectOperation.execute(ParallelSelectOperation.java:61)
	at org.eclipse.epsilon.eol.execute.operations.declarative.ExistsOperation.execute(ExistsOperation.java:24)
	at org.eclipse.epsilon.eol.execute.operations.declarative.ExistsOperation.execute(ExistsOperation.java:1)
	at org.eclipse.epsilon.eol.dom.FirstOrderOperationCallExpression.execute(FirstOrderOperationCallExpression.java:118)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.executeImpl(ExecutorFactory.java:204)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.execute(ExecutorFactory.java:233)
	at org.eclipse.epsilon.eol.dom.ReturnStatement.execute(ReturnStatement.java:39)
	at org.eclipse.epsilon.eol.dom.ReturnStatement.execute(ReturnStatement.java:1)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.executeImpl(ExecutorFactory.java:204)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.execute(ExecutorFactory.java:233)
	at org.eclipse.epsilon.eol.dom.StatementBlock.execute(StatementBlock.java:68)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.executeImpl(ExecutorFactory.java:204)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.execute(ExecutorFactory.java:233)
	at org.eclipse.epsilon.eol.dom.Operation.executeBody(Operation.java:191)
	at org.eclipse.epsilon.eol.dom.Operation.execute(Operation.java:173)
	at org.eclipse.epsilon.eol.dom.Operation.execute(Operation.java:148)
	at org.eclipse.epsilon.eol.dom.OperationCallExpression.execute(OperationCallExpression.java:157)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.executeImpl(ExecutorFactory.java:204)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.execute(ExecutorFactory.java:233)
	at org.eclipse.epsilon.eol.dom.AssignmentStatement.execute(AssignmentStatement.java:97)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.executeImpl(ExecutorFactory.java:204)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.execute(ExecutorFactory.java:233)
	at org.eclipse.epsilon.eol.dom.StatementBlock.execute(StatementBlock.java:68)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.executeImpl(ExecutorFactory.java:204)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.execute(ExecutorFactory.java:233)
	at org.eclipse.epsilon.eol.dom.ExecutableBlock.executeBlockOrExpressionAst(ExecutableBlock.java:103)
	at org.eclipse.epsilon.eol.dom.ExecutableBlock.executeBody(ExecutableBlock.java:177)
	at org.eclipse.epsilon.eol.dom.ExecutableBlock.execute(ExecutableBlock.java:128)
	at org.eclipse.epsilon.eol.dom.ExecutableBlock.execute(ExecutableBlock.java:181)
	at org.eclipse.epsilon.evl.dom.Constraint.executeCheckBlock(Constraint.java:153)
	at org.eclipse.epsilon.evl.dom.Constraint.check(Constraint.java:123)
	at org.eclipse.epsilon.evl.dom.Constraint.execute(Constraint.java:80)
	at org.eclipse.epsilon.evl.dom.Constraint.execute(Constraint.java:1)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.executeImpl(ExecutorFactory.java:222)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.execute(ExecutorFactory.java:263)
	at org.eclipse.epsilon.evl.dom.ConstraintContext.execute(ConstraintContext.java:201)
	at org.eclipse.epsilon.evl.EvlModule.checkConstraints(EvlModule.java:278)
	at org.eclipse.epsilon.evl.EvlModule.processRules(EvlModule.java:266)
	at org.eclipse.epsilon.evl.EvlModule.processRules(EvlModule.java:1)
	at org.eclipse.epsilon.erl.ErlModule.executeImpl(ErlModule.java:174)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.executeImpl(ExecutorFactory.java:207)
	at org.eclipse.epsilon.eol.execute.ExecutorFactory.execute(ExecutorFactory.java:233)
	at org.eclipse.epsilon.eol.EolModule.execute(EolModule.java:476)
	at org.eclipse.epsilon.evl.EvlModule.execute(EvlModule.java:300)
	at org.eclipse.epsilon.evl.EvlModule.execute(EvlModule.java:1)
	at org.eclipse.epsilon.eol.launch.EolRunConfiguration.execute(EolRunConfiguration.java:195)
	at org.eclipse.epsilon.evl.launch.EvlRunConfiguration.execute(EvlRunConfiguration.java:128)
	at org.eclipse.epsilon.evl.launch.EvlRunConfiguration.execute(EvlRunConfiguration.java:1)
	at org.eclipse.epsilon.common.launch.ProfilableRunConfiguration.call(ProfilableRunConfiguration.java:234)
	at org.eclipse.epsilon.common.launch.ProfilableRunConfiguration.run(ProfilableRunConfiguration.java:265)
	at org.eclipse.epsilon.eol.engine.test.acceptance.util.EolEquivalenceTests.setUpEquivalenceTest(EolEquivalenceTests.java:106)
	at org.eclipse.epsilon.evl.engine.test.acceptance.equivalence.EvlParallelOperationsTests.setUpBeforeClass(EvlParallelOperationsTests.java:88)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:757)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
```

This is happening at test setup time - using `jstack`, it appears this is creating 4000+ threads before any tests run and without disposing any previously used threads, which is hitting the OS limits set in recent versions of MacOS for Apple Silicon.

The test suite should be restructured so it doesn't spin up all these threads in advance, but rather run EVL programs, obtain their results, and then shut down those threads.

**Comment by agarciadom** [comment]
Turns out that we did have a thread leak inside Epsilon's parallel tasks. When we begin a parallel task and start a new EolContextParallel in a non-"short circuiting" piece of work (basically, one we cannot interrupt), we weren't shutting down the executor appropriately (which means the threads were never being freed after everything was done).

The fix is to always shut down the executor - if we must interrupt then we use `shutdownNow` as before (which sends an interrupt to all tasks), and if we shouldn't interrupt then we just do `shutdown` (which tells the Executor we won't be scheduling any more tasks, so it can free up the threads once they are all done).

I'll submit a PR with the fix, as well as a rewrite of the module equivalence tests in Epsilon.
